cdap-ui/app/cdap/components/Market/index.js

```diff
@@ -51,55 +49,147 @@ export default class Market extends Component {
-        }
-        tabConfig.tabs = tabConfig.tabs.concat(missingTabsFromConfig);
+  }
 
-        this.setState({
-          tabsList,
-          tabConfig
+  componentWillUnmount() {
+    MarketStore.dispatch({ type: 'RESET' });
+
+    if (this.sub) {
+      this.sub();
+    }
+  }
+
+  getCategories = (packages) => {
+    MyMarketApi.getCategories()
+      .subscribe((categories) => {
+        this.processPackagesAndCategories(packages, categories);
+      }, () => {
+        // If categories do not come from backend, revert back to get categories from existing packages
+        const categoriesMap = {};
+        packages.forEach((pack) => {
+          pack.categories.forEach((category) => {
+            categoriesMap[category] = true;
+          });
+        });
+
+        let aggregateCategories = [];
+
+        DEFAULT_CATEGORIES.forEach((cat) => {
+          if (categoriesMap[cat]) {
+            aggregateCategories.push(cat);
+            delete categoriesMap[cat];
+          }
         });
+
+        const remainingCategories = Object.keys(categoriesMap);
+
+        aggregateCategories = aggregateCategories
+          .concat(remainingCategories)
+          .map((cat) => {
+            return {
+              name: cat,
+              hasIcon: false
+            };
+          });
+
+        this.processPackagesAndCategories(packages, aggregateCategories);
       });
   }
 
+  processPackagesAndCategories(packages, categories) {
+    const newState = {
+      tabConfig: this.constructTabConfig(categories)
+    };
+    const searchFilter = find(newState.tabConfig.tabs, { filter: MarketStore.getState().filter });
+
+    if (searchFilter) {
+      newState.activeTab = searchFilter.id;
+    }
+
+    this.setState(newState);
+    MarketAction.setList(packages);
+  }
+
+  constructTabConfig(categories) {
+    const tabConfig = {
+      defaultTab: 1,
+      defaultTabContent: <AllTabContents />,
+      layout: 'vertical',
+    };
+
+    const tabs = [
+      {
+        id: 1,
+        filter: '*',
+        icon: {
+          type: 'font-icon',
+          arguments: {
+            data: 'icon-all'
+          }
+        },
+        name: T.translate('features.Market.tabs.all'),
+        content: <AllTabContents />
+      }
+    ];
+
+    categories.forEach((category) => {
+      const categoryContent = CATEGORY_MAP[category.name] || {};
+      const name = categoryContent.displayName || category.name;
+
+      let icon;
+
+      if (category.hasIcon) {
+        icon = {
+          type: 'link',
+          arguments: {
+            url: MyMarketApi.getCategoryIcon(category.name)
+          }
+        };
+      } else if (categoryContent.displayName) {
+        icon = {
+          type: 'font-icon',
+          arguments: {
+            data: categoryContent.icon
+          }
+        };
+      } else {
+        icon = {
+          type: 'font-icon',
+          arguments: {
+            data: `icon-${category.name[0].toUpperCase()}`
```

Comment:
nit: Just to be safe, can you check if category has name and can query at index `0`? Would avoid a js error.